### ISSUE TITLE
reading file twice (in PutObject)

### ISSUE THREAD

**Issue by crusader-mike** [aws/aws-sdk-cpp#711]
Apparently AWS SDK needs to calculate some sort of signature before sending bytes over wire (in PutObject/etc) -- related stream gets read and then rewound to start and read again. Is there any way to avoid this double "read file from disk" besides using streambuf that keeps entire file in memory?

**Comment by JonathanHenson** [comment]
The default ctor args for s3 turn the payload hashing off of you are using tls. Also, setting the content length field will prevent the seek operations to figure it out. By controlling both of these settings, the file will only be read once.

**Comment by crusader-mike** [comment]
Huh... I am using https and reading file (in my own implementation of streambuf) only in `underflow()` -- according to your criteria SDK shouldn't scan my file twice. I'll double check tonight

**Comment by crusader-mike** [comment]
After examining my (relatively recent) SDK sources I came to a conclusion that it is impossible to switch off "payload signing" without overriding request class methods -- all requests get payload signed if `<request-class>::SignBody()` returns true. Similar flag passed by S3Client into aws signer class gets used only with payload-less requests -- i.e. not used at all. The only way to avoid payload being signed is to derive from requess-class, override `SignBody()` and return false.

I dug in this a bit more and found that Amazon S3 (as of now) allows unsigned payloads, but you can configure your bucket to deny such requests.  Is it a prelude to total ban on unsigned payloads service wide?

In any case -- it would be very nice to have a switch in S3Client (or library-wide?) that would forcefully disable/enable payload signing (regardless of HTTP/HTTPS flag). Please :)

Until then I believe I have only these options:
1. live with 'double read'
2. load entire file into memory -- not end of the world, since I use multipart upload for large files anyway
3. override SignBody() of each request I use and hope users don't use HTTP urls in my program

Sounds like option \#2 is the least of all evil right now.

**Comment by marcomagdy** [comment]
This is a bug. We need to fix it.

**Comment by crusader-mike** [comment]
Thank you. Would be nice to be able to skip payload signature in HTTP case too...

**Comment by crusader-mike** [comment]
A note about option N2 -- multipart upload is limited to 10'000 chunks. Which means for files bigger than 50GB  chunk size will start growing and at some point they won't fit into memory (especially if you upload them in parallel). I.e. you'll have to limit memory buffer and you'll end up reading entire file twice.

**Comment by wps132230** [comment]
Hi, you can avoid reading file twice by skipping payload signature with the latest update(ver 1.3.0) now:
```c++
auto s3Client = Aws::MakeShared<S3Client>(ALLOCATION_TAG, credentialsProvider, config, AWSAuthV4Signer::PayloadSigningPolicy::Never /*signPayloads*/, true /*useVirtualAddressing*/);
```
in which `AWSAuthV4Signer::PayloadSigningPolicy::Never` will overrule `SignBody()`, even though it's `true`, then the payload will never be signed.

Please consult this commit for more details: https://github.com/aws/aws-sdk-cpp/commit/b9ccf4d0ded1dccd1fe66fbba6b5c9c306fbb249

**Comment by crusader-mike** [comment]
Thanks, Pushen!
![image](https://user-images.githubusercontent.com/22357016/32641613-999b32dc-c594-11e7-8729-360209930b1d.png)

**Comment by crusader-mike** [comment]
@singku @wps132230 I've discovered that with HTTP endpoint my source files/objects are still being read twice. 

Culprit is check for `request.GetUri().GetScheme() != Http::Scheme::HTTPS` check at [this line](https://github.com/aws/aws-sdk-cpp/blob/b24dc7bcf286078083bad68f7a13f27457fb0073/aws-cpp-sdk-core/source/auth/AWSAuthSigner.cpp#L202).

Can someone explain why SDK insists on calculating payload hash on non-HTTPS endpoints even if I explicitly asked it not to "sign boody"?

Is there an easy way to switch it off without substituting request signer object?
